File: 3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.ts
function lengthOfLongestSubstring(s: string): number {
    let left = 0;
    let maxLength = 0;
    const seenChars = new Set<string>();

    for (let right = 0; right < s.length; right++) {
        // If the character is already in the set, shrink the window by moving the left pointer
        while (seenChars.has(s[right])) {
            seenChars.delete(s[left]);
            left++;
        }
        
        // Add the current character to the set
        seenChars.add(s[right]);

        // Update the maximum length of the substring found
        maxLength = Math.max(maxLength, right - left + 1);
    }

    return maxLength;
}

File: 3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.go
func lengthOfLongestSubstring(s string) int {
    n := len(s)
    if n <= 1 {
        return n
    }
    
    // Use a map to store the last index of each character
    lastIndex := make(map[byte]int)
    
    // Initialize the sliding window
    start, maxLen := 0, 0
    
    for i := 0; i < n; i++ {
        // If the current character has been seen before and its last index is within the current window
        if index, ok := lastIndex[s[i]]; ok && index >= start {
            // Move the start of the window to the position after the last occurrence of the character
            start = index + 1
        }
        
        // Update the last index of the current character
        lastIndex[s[i]] = i
        
        // Calculate the length of the current window
        currLen := i - start + 1
        
        // Update the maximum length if necessary
        if currLen > maxLen {
            maxLen = currLen
        }
    }
    
    return maxLen
}